pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,219 @@
+package registry
+
+import (
+	"errors"
+	"math"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/btree"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/origin/pkg/oauth/apis/oauth"
+	oauthclient "github.com/openshift/origin/pkg/oauth/generated/internalclientset/typed/oauth/internalversion"
+)
+
+var ErrTimedout = errors.New("Token timed out")
+
+type tokenData struct {
+	name    string
+	client  string
+	seen    time.Time
+	timeout time.Time
+}
+
+type tokenDataRef struct {
+	name    string
+	timeout time.Time
+}
+
+func (a *tokenDataRef) Less(than btree.Item) bool {
+
+	if a.timeout.Equal(than.(*tokenDataRef).timeout) {
+		return a.name < than.(*tokenDataRef).name
+	}
+	return a.timeout.Before(than.(*tokenDataRef).timeout)
+}
+
+type TokenTimeoutUpdater struct {
+	oauthClient    oauthclient.OAuthClientInterface
+	clientTimeouts map[string]int32
+	tokens         oauthclient.OAuthAccessTokenInterface
+	tokenChannel   chan tokenData
+	data           map[string]tokenData
+	tree           *btree.BTree
+}
+
+func NewTokenTimeoutUpdater(tokens oauthclient.OAuthAccessTokenInterface, oauthClient oauthclient.OAuthClientInterface) *TokenTimeoutUpdater {
+	ttu := &TokenTimeoutUpdater{
+		oauthClient,
+		make(map[string]int32),
+		tokens,
+		make(chan tokenData, 100),
+		make(map[string]tokenData),
+		// FIXME: what is the right degree for the btree
+		btree.New(32),
+	}
+	go ttu.Start()
+	return ttu
+}
+
+// CheckTimeout is called with a token when it is seen by an authenticator
+// it touches only the tokenChannel so it is safe to call from other threads
+func (a *TokenTimeoutUpdater) CheckTimeout(name string, token *oauth.OAuthAccessToken) error {
+	var timeout time.Time
+	timenow := time.Now()
+	if token.TimeoutsIn > 0 {
+		timeout = token.CreationTimestamp.Time.Add(time.Duration(token.TimeoutsIn) * time.Second)
+		if timeout.Before(timenow) {
+			return ErrTimedout
+		}
+	}
+
+	// After a positive timeout check we need to update the timeout and
+	// schedule an update so that we can either set or update the Timeout
+	a.tokenChannel <- tokenData{name, token.ClientName, timenow, timeout}
+	return nil
+}
+
+// updates known clients timeouts, removing clients if they go away
+// and calculate the flush period based on actual timeouts, with a
+// 10 minutes default
+func (a *TokenTimeoutUpdater) updateClientTimeouts() time.Duration {
```

Comment:
Either encode the client timeout in the token or use a shared informer for OAuth clients.  Do not build your own.

Reply:
Note that this is not just caching, but I guess I can plumb in a watcher for OAuth Clients